server/schedule/operator.go

```diff
@@ -221,6 +221,46 @@ func (o *Operator) Influence(opInfluence OpInfluence, region *core.RegionInfo) {
 	}
 }
 
+// OperatorHistory is used to log and visualize completed operators.
+type OperatorHistory struct {
+	FinishTime time.Time
+	From, To   uint64
+	Kind       core.ResourceKind
+}
+
+// History transfers the operator's steps to operator histories.
+func (o *Operator) History() []OperatorHistory {
+	now := time.Now()
+	var histories []OperatorHistory
+	var addPeerStores, removePeerStores []uint64
+	for _, step := range o.steps {
+		switch s := step.(type) {
+		case TransferLeader:
+			histories = append(histories, OperatorHistory{
+				FinishTime: now,
+				From:       s.FromStore,
+				To:         s.ToStore,
+				Kind:       core.LeaderKind,
+			})
+		case AddPeer:
+			addPeerStores = append(addPeerStores, s.ToStore)
+		case RemovePeer:
+			removePeerStores = append(removePeerStores, s.FromStore)
+		}
+	}
+	for i := range addPeerStores {
+		if i < len(removePeerStores) {
+			histories = append(histories, OperatorHistory{
+				FinishTime: now,
+				From:       removePeerStores[i],
+				To:         addPeerStores[i],
```

Comment:
I think there is no order guarantee.

Reply:
Yes, in fact we do not care about the order, only the changing trend.